dependencies.yaml

```
@@ -180,11 +198,55 @@ dependencies:
       - output_types: [conda, requirements, pyproject]
         packages:
           - cloudpickle
-          - cudf==24.2.*
-          - cupy
           - dask
           - distributed
           - numba>=0.57.1
           - pytest
           - pytest-asyncio
           - pytest-rerunfailures
+  depends_on_cupy:
+    common:
+      - output_types: conda
+        packages:
+          - cupy>=12.0.0
+    specific:
+      - output_types: [requirements, pyproject]
+        matrices:
+          # All CUDA 12 + x86_64 versions
+          - matrix: {cuda: "12.*", arch: x86_64}
+            packages:
+              - cupy-cuda12x>=12.0.0
+          # All CUDA 12 + aarch64 versions
+          - matrix: {cuda: "12.*", arch: aarch64}
+            packages:
+              - cupy-cuda12x -f https://pip.cupy.dev/aarch64 # TODO: Verify that this works.
+          # All CUDA 11 + x86_64 versions
+          - matrix: {cuda: "11.*", arch: x86_64}
+            packages:
+              - cupy-cuda11x>=12.0.0
+          # All CUDA 11 + aarch64 versions
+          - matrix: {cuda: "11.*", arch: aarch64}
+            packages:
+              - cupy-cuda11x -f https://pip.cupy.dev/aarch64 # TODO: Verify that this works.
+          - {matrix: null, packages: [cupy-cuda11x>=12.0.0]}
```

cupy has ARM wheels on PyPI now and no special index is needed. We can simplify this.

FYI, I've already applied this change in several of the PRs testing CUDA 12.2, at the same time that I updated and simplified the matrices for other RAPIDS packages.

```suggestion
      - output_types: [requirements, pyproject]
        matrices:
          - matrix: {cuda: "12.*"}
            packages:
              - cupy-cuda12x>=12.0.0
          - matrix: {cuda: "11.*"}
            packages:
              - cupy-cuda11x>=12.0.0
          - {matrix: null, packages: [cupy-cuda11x>=12.0.0]}
```